pkg/crcontrollers/storagegroup/storagegroupcontroller.go

```diff
@@ -0,0 +1,586 @@
+package storagegroup
+
+import (
+	"context"
+	"fmt"
+	"reflect"
+	"strconv"
+	"time"
+
+	"github.com/sirupsen/logrus"
+	k8serrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/runtime"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/event"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+
+	api "github.com/dell/csi-baremetal/api/generated/v1"
+	apiV1 "github.com/dell/csi-baremetal/api/v1"
+	accrd "github.com/dell/csi-baremetal/api/v1/availablecapacitycrd"
+	"github.com/dell/csi-baremetal/api/v1/drivecrd"
+	sgcrd "github.com/dell/csi-baremetal/api/v1/storagegroupcrd"
+	"github.com/dell/csi-baremetal/pkg/base/k8s"
+	"github.com/dell/csi-baremetal/pkg/base/util"
+)
+
+const (
+	sgFinalizer               = "dell.emc.csi/sg-cleanup"
+	sgTempStatusAnnotationKey = "storagegroup.csi-baremetal.dell.com/status"
+	contextTimeoutSeconds     = 60
+)
+
+// Controller to reconcile storagegroup custom resource
+type Controller struct {
+	client         *k8s.KubeClient
+	log            *logrus.Entry
+	crHelper       k8s.CRHelper
+	cachedCrHelper k8s.CRHelper
+}
+
+// NewController creates new instance of Controller structure
+// Receives an instance of base.KubeClient and logrus logger
+// Returns an instance of Controller
+func NewController(client *k8s.KubeClient, k8sCache k8s.CRReader, log *logrus.Logger) *Controller {
+	c := &Controller{
+		client:         client,
+		crHelper:       k8s.NewCRHelperImpl(client, log),
+		cachedCrHelper: k8s.NewCRHelperImpl(client, log).SetReader(k8sCache),
+		log:            log.WithField("component", "StorageGroupController"),
+	}
+	return c
+}
+
+// SetupWithManager registers Controller to ControllerManager
+func (c *Controller) SetupWithManager(mgr ctrl.Manager) error {
+	return ctrl.NewControllerManagedBy(mgr).
+		For(&sgcrd.StorageGroup{}).
+		WithOptions(controller.Options{}).
+		Watches(&source.Kind{Type: &drivecrd.Drive{}}, &handler.EnqueueRequestForObject{}).
+		WithEventFilter(predicate.Funcs{
+			UpdateFunc: func(e event.UpdateEvent) bool {
+				return c.filterUpdateEvent(e.ObjectOld, e.ObjectNew)
+			},
+		}).
+		Complete(c)
+}
+
+func (c *Controller) filterUpdateEvent(old runtime.Object, new runtime.Object) bool {
+	var (
+		oldDrive *drivecrd.Drive
+		newDrive *drivecrd.Drive
+		ok       bool
+	)
+	if oldDrive, ok = old.(*drivecrd.Drive); !ok {
```

Comment:
I don't think this line code is necessary here.  we can remove it.

Reply:
Hi Libin, at present, the storagegroup controller watches the events of both drive objects and storagegroup objects, so this code here is to distinguish the non-drive object, i.e. storagegroup object. Without this line, we will miss the storagegroup update event and then the storage group reconcilation wouldn't be normally handled. Though the StorageGroupSpec update is not supported, we actually still do the necessary update on the whole StorageGroup object during the storagegroup reconcilation for storagegroup status tracking, finalizer addition or removal, node addtion support etc. Therefore, the catch of storagegroup object update event is still required here.
Here, I just use the similar code logic to those already used in other CR controllers watching object events of multiple types.

Reply:
yes, I originally thought you can "return true" in the end(Line 84)  to handle other case.  anyway,  this is ok.

Reply:
@libzhang Very sorry for misunderstanding of your original comments. 
Your suggested code change is much better if I want to catch all storagegroup update events and only storagegroup label update event for drive objects.
I will change the code here accordingly.